### DISCUSSION TITLE
how should i export all info

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1467] [Q&A]
If different st files have cross-reference ,how should i export all  related info,export format can be json or other tree structure
for example aa.st define a property 'param1',which refer struct 'Panda',struct Panda have 'type' property ,which refer struct 'Animal',
how can i  return all related property by  property 'param1'
![image](https://github.com/eclipse-langium/langium/assets/26995040/2ef1d825-c1d0-49cb-8549-d97c12b6605a)
![image](https://github.com/eclipse-langium/langium/assets/26995040/57a4917d-ec32-4498-b594-d5e15a23718f)

**Comment by msujew** [comment]
Hey @zrwsmd,

we usually just store everything in maps (or use the `DocumentCache` or `WorkspaceCache`). See [this type system inference](https://github.com/TypeFox/langium-lox/blob/a821a08b967e1eb260079ca6e4a612a1c46000e9/src/language-server/type-system/infer.ts#L5) method that uses a cache to associate type information to a given AST node.

**Reply by zrwsmd** [reply]
i mean export all current workspace file info(can be json format for example) to other people  by vscode extension,so that other people can get this info.for example,if i have two st files at my workspace (main2.st and functionblock2.st)
i want to export these var declarations info like below json format
[{
		"fileName": "D://xx/projectName/workspace/main2.st",
		"elements": [{
				"varName": "state",
				"type": "INT"
			},
			{
				"varName": "tempConverter",
				"type": "TemperatureConverter2",
				"refFile": "D://xx/projectName/workspace/functionblock2.st@TemperatureConverter2"
			}
		]
	},
	{
		"fileName": "D://xx/projectName/workspace/functionblock2.st",
		"elements": [{
			"varName": "state",
			"type": "R_TRIG"
		}]
	}
]

![image](https://github.com/eclipse-langium/langium/assets/26995040/3f7d1bf9-0da2-4984-9f11-813ff068b97f)

**Reply by zrwsmd** [reply]
similar to this precomputedScopes data,but how should i expose it through vscode extension so that my colleague can get this info
(we communicate data info through vscode extension)
![image](https://github.com/eclipse-langium/langium/assets/26995040/61321b4e-57ec-4136-ab9c-331f9f9c2c83)
![image](https://github.com/eclipse-langium/langium/assets/26995040/a90b02e4-b8ee-4de5-9bb4-a3c9d6865e06)

**Reply by cdietrich** [reply]
You might look into ExecuteCommandHandler extension point in combination with LangiumDocuments api to access all documents

**Reply by zrwsmd** [reply]
is there any example or document to refer?

**Reply by cdietrich** [reply]
I don’t know simply subclass and bind AbstractExecuteCommandHandler

maybe you can find someone else using it

**Reply by msujew** [reply]
You can probably use a similar mechanism that we use internally to display the railroad diagrams in the vscode extension. See [here](https://github.com/eclipse-langium/langium/blob/main/packages/langium-vscode/src/language-server/railroad-handler.ts). Also you might want to use the `services.serializer.JsonSerializer` to serialize/deserialize an AST to send across the wire.
